api4/user.go

```diff
@@ -1131,6 +1151,37 @@ func updateUser(c *Context, w http.ResponseWriter, r *http.Request) {
 	w.Write([]byte(ruser.ToJson()))
 }
 
+// checkProviderAttributes returns the empty string if the patch can be applied without
+// overriding attributes set by the user's login provider; otherwise, the name of the offending
+// field is returned.
+func checkProviderAttributes(a app.AppIface, ouser *model.User, patch *model.UserPatch) string {
+	tryingToChange := func(patchValue *string, userValue *string) bool {
+		return patchValue != nil && *patchValue != *userValue
+	}
+	LdapSettings := &a.Config().LdapSettings
+	SamlSettings := &a.Config().SamlSettings
+
+	conflictField := ""
+	if a.Ldap() != nil &&
+		(ouser.IsLDAPUser() || (ouser.IsSAMLUser() && *SamlSettings.EnableSyncWithLdap)) {
+		conflictField = a.Ldap().CheckProviderAttributes(LdapSettings, ouser, patch)
+	} else if a.Saml() != nil && ouser.IsSAMLUser() {
+		conflictField = a.Saml().CheckProviderAttributes(SamlSettings, ouser, patch)
+	} else if ouser.IsOAuthUser() {
+		if tryingToChange(patch.FirstName, &ouser.FirstName) ||
+			tryingToChange(patch.LastName, &ouser.LastName) {
+			conflictField = "full name"
+		}
+	}
+	// If any login provider is used, then the username may not be changed
+	if ouser.AuthService != "" {
+		if tryingToChange(patch.Username, &ouser.Username) && conflictField == "" {
+			conflictField = "username"
+		}
+	}
```

Comment:
I think this can be move to the top of the function right before you check if ldap is enabled and etc. I would say you can move there and do something like this
```suggestion
	// If any login provider is used, then the username may not be changed
	if ouser.AuthService != "" && tryingToChange(patch.Username, &ouser.Username) {
	        return "username"
	}
```
It makes it clear to say something like "If you're using any auth service and you are trying to change username, we will return since you're not allowed and not check for specifics cause the rest of the code doesn't matter" Hopefully that made sense. I feel this order will make it more clear from a developers standpoint.

Reply:
Good idea. I like this a lot better.